File: backend/src/AgentTeam/Services/TeamRepository.php
<?php

declare(strict_types=1);

namespace AgentTeam\Services;

use AgentTeam\Models\Team;
use PDO;

/**
 * Team Repository
 *
 * Handles CRUD operations for teams.
 *
 * @see docs/agentDesign.md
 */
class TeamRepository
{
    private PDO $db;
    private AgentRepository $agentRepository;

    public function __construct(PDO $db, ?AgentRepository $agentRepository = null)
    {
        $this->db = $db;
        $this->agentRepository = $agentRepository ?? new AgentRepository($db);
    }

    /**
     * Find a team by ID
     */
    public function findById(int $id): ?Team
    {
        $stmt = $this->db->prepare("SELECT * FROM agent_teams WHERE id = ?");
        $stmt->execute([$id]);
        $row = $stmt->fetch(PDO::FETCH_ASSOC);

        return $row ? new Team($row) : null;
    }

    /**
     * Find a team by ID with its agents (includes pipeline info)
     */
    public function findByIdWithAgents(int $id): ?Team
    {
        $team = $this->findById($id);
        if (!$team) {
            return null;
        }

        // Get agents with pipeline position info
        $agentsWithInfo = $this->agentRepository->findByTeamIdWithPipelineInfo($id);
        $team->setAgentsWithPipelineInfo($agentsWithInfo);

        return $team;
    }

    /**
     * Find all teams for a user
     */
    public function findByUser(int $userId): array
    {
        $stmt = $this->db->prepare(
            "SELECT * FROM agent_teams WHERE user_id = ? ORDER BY name ASC"
        );
        $stmt->execute([$userId]);

        return array_map(
            fn($row) => new Team($row),
            $stmt->fetchAll(PDO::FETCH_ASSOC)
        );
    }

    /**
     * Find all teams for a user with their agents (includes pipeline info)
     */
    public function findByUserWithAgents(int $userId): array
    {
        $teams = $this->findByUser($userId);

        foreach ($teams as $team) {
            // Get agents with pipeline position info
            $agentsWithInfo = $this->agentRepository->findByTeamIdWithPipelineInfo($team->getId());
            $team->setAgentsWithPipelineInfo($agentsWithInfo);
        }

        return $teams;
    }

    /**
     * Create a new team
     */
    public function create(Team $team): Team
    {
        $sql = "INSERT INTO agent_teams (user_id, workspace_id, name, description)
                VALUES (:user_id, :workspace_id, :name, :description)";

        $data = $team->toArray();

        $stmt = $this->db->prepare($sql);
        $stmt->execute([
            'user_id' => $data['user_id'],
            'workspace_id' => $data['workspace_id'],
            'name' => $data['name'],
            'description' => $data['description'],
        ]);

        $id = (int) $this->db->lastInsertId();
        return $this->findById($id);
    }

    /**
     * Update an existing team
     */
    public function update(Team $team): Team
    {
        $sql = "UPDATE agent_teams SET
                name = :name,
                description = :description,
                workspace_id = :workspace_id
                WHERE id = :id";

        $data = $team->toArray();

        $stmt = $this->db->prepare($sql);
        $stmt->execute([
            'id' => $data['id'],
            'name' => $data['name'],
            'description' => $data['description'],
            'workspace_id' => $data['workspace_id'],
        ]);

        return $this->findById($team->getId());
    }

    /**
     * Delete a team and all its agents
     */
    public function delete(int $id): bool
    {
        // First, delete all agents in this team
        $stmt = $this->db->prepare("DELETE FROM agents WHERE team_id = ?");
        $stmt->execute([$id]);

        // Then delete the team
        $stmt = $this->db->prepare("DELETE FROM agent_teams WHERE id = ?");
        return $stmt->execute([$id]);
    }

    /**
     * Check if a user can access a team
     */
    public function canUserAccess(int $userId, int $teamId): bool
    {
        $team = $this->findById($teamId);
        if (!$team) {
            return false;
        }

        // Owner always has access
        return $team->getUserId() === $userId;
    }

    /**
     * Check if a user owns a team
     */
    public function isOwner(int $userId, int $teamId): bool
    {
        $team = $this->findById($teamId);
        return $team && $team->getUserId() === $userId;
    }

    /**
     * Count teams for a user
     */
    public function countByUser(int $userId): int
    {
        $stmt = $this->db->prepare(
            "SELECT COUNT(*) FROM agent_teams WHERE user_id = ?"
        );
        $stmt->execute([$userId]);
        return (int) $stmt->fetchColumn();
    }

    /**
     * Count agents in a team
     */
    public function countAgents(int $teamId): int
    {
        $stmt = $this->db->prepare(
            "SELECT COUNT(*) FROM agents WHERE team_id = ? AND enabled = 1"
        );
        $stmt->execute([$teamId]);
        return (int) $stmt->fetchColumn();
    }
}

File: backend_python/app/agent_team/services/team_repository.py
"""Port of backend/src/AgentTeam/Services/TeamRepository.php (195 lines).

Handles CRUD operations for teams.
"""
from __future__ import annotations

from app.agent_team.models.team import Team
from app.agent_team.services.agent_repository import AgentRepository
from app.support.phpcompat import php_intval


class TeamRepository:
    def __init__(self, db, agent_repository: AgentRepository | None = None):
        self.db = db
        self.agentRepository = agent_repository if agent_repository is not None else AgentRepository(db)

    def findById(self, id: int) -> Team | None:
        """PHP 31-38."""
        row = self.db.fetch_one("SELECT * FROM agent_teams WHERE id = ?", [id])
        return Team(row) if row else None

    def findByIdWithAgents(self, id: int) -> Team | None:
        """PHP 43-55."""
        team = self.findById(id)
        if not team:
            return None

        # Get agents with pipeline position info
        agents_with_info = self.agentRepository.findByTeamIdWithPipelineInfo(id)
        team.setAgentsWithPipelineInfo(agents_with_info)

        return team

    def findByUser(self, user_id: int) -> list[Team]:
        """PHP 60-71."""
        rows = self.db.fetch_all(
            "SELECT * FROM agent_teams WHERE user_id = ? ORDER BY name ASC", [user_id]
        )
        return [Team(row) for row in rows]

    def findByUserWithAgents(self, user_id: int) -> list[Team]:
        """PHP 76-87."""
        teams = self.findByUser(user_id)

        for team in teams:
            # Get agents with pipeline position info
            agents_with_info = self.agentRepository.findByTeamIdWithPipelineInfo(team.getId())
            team.setAgentsWithPipelineInfo(agents_with_info)

        return teams

    def create(self, team: Team) -> Team:
        """PHP 92-109."""
        sql = (
            "INSERT INTO agent_teams (user_id, workspace_id, name, description)\n"
            "                VALUES (:user_id, :workspace_id, :name, :description)"
        )

        data = team.toArray()

        new_id = self.db.insert(sql, {
            'user_id': data['user_id'],
            'workspace_id': data['workspace_id'],
            'name': data['name'],
            'description': data['description'],
        })

        return self.findById(new_id)

    def update(self, team: Team) -> Team:
        """PHP 114-133."""
        sql = (
            "UPDATE agent_teams SET\n"
            "                name = :name,\n"
            "                description = :description,\n"
            "                workspace_id = :workspace_id\n"
            "                WHERE id = :id"
        )

        data = team.toArray()

        self.db.execute(sql, {
            'id': data['id'],
            'name': data['name'],
            'description': data['description'],
            'workspace_id': data['workspace_id'],
        })

        return self.findById(team.getId())

    def delete(self, id: int) -> bool:
        """PHP 138-147. Deletes all agents in the team first, then the team.
        `$stmt->execute([$id])` is the boolean success of the exec, not a
        rowCount check."""
        # First, delete all agents in this team
        self.db.execute("DELETE FROM agents WHERE team_id = ?", [id])

        # Then delete the team
        self.db.execute("DELETE FROM agent_teams WHERE id = ?", [id])
        return True

    def canUserAccess(self, user_id: int, team_id: int) -> bool:
        """PHP 152-161."""
        team = self.findById(team_id)
        if not team:
            return False

        # Owner always has access
        return team.getUserId() == user_id

    def isOwner(self, user_id: int, team_id: int) -> bool:
        """PHP 166-170."""
        team = self.findById(team_id)
        return bool(team) and team.getUserId() == user_id

    def countByUser(self, user_id: int) -> int:
        """PHP 175-182."""
        col = self.db.fetch_column("SELECT COUNT(*) FROM agent_teams WHERE user_id = ?", [user_id])
        return php_intval(col[0]) if col else 0

    def countAgents(self, team_id: int) -> int:
        """PHP 187-194."""
        col = self.db.fetch_column(
            "SELECT COUNT(*) FROM agents WHERE team_id = ? AND enabled = 1", [team_id]
        )
        return php_intval(col[0]) if col else 0
